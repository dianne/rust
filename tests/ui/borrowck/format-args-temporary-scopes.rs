//! Test for #145784 as it relates to format arguments: arguments to macros such as `println!`
//! should obey normal temporary scoping rules.
//@ revisions: e2021 e2024
//@ [e2021] edition: 2021
//@ [e2024] edition: 2024

fn temp() {}

fn main() {
    // In Rust 2024, block tail expressions are temporary scopes, so the result of `temp()` is
    // dropped after evaluating `&temp()`.
    println!("{:?}", { &temp() });
    //[e2024]~^ ERROR: temporary value dropped while borrowed [E0716]

    // In Rust 1.89, `format_args!` extended the lifetime of all extending expressions in its
    // arguments when provided with two or more arguments. This caused the result of `temp()` to
    // outlive the result of the block, making this compile.
    println!("{:?}{:?}", { &temp() }, ());
    //[e2024]~^ ERROR: temporary value dropped while borrowed [E0716]

    // In real-world projects, this typically appeared in `if` expressions with a `&str` in one
    // branch and a reference to a `String` temporary in the other. Since the consequent and `else`
    // blocks of `if` expressions are temporary scopes in all editions, this affects Rust 2021 and
    // earlier as well.
    println!("{:?}{:?}", (), if true { &format!("") } else { "" });
    //~^ ERROR: temporary value dropped while borrowed [E0716]
}
