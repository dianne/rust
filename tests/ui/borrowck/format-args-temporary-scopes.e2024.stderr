error[E0716]: temporary value dropped while borrowed
  --> $DIR/format-args-temporary-scopes.rs:12:25
   |
LL |     println!("{:?}", { &temp() });
   |                      ---^^^^^---
   |                      |  |    |
   |                      |  |    temporary value is freed at the end of this statement
   |                      |  creates a temporary value which is freed while still in use
   |                      borrow later used here
   |
   = note: consider using a `let` binding to create a longer lived value

error[E0716]: temporary value dropped while borrowed
  --> $DIR/format-args-temporary-scopes.rs:18:29
   |
LL |     println!("{:?}{:?}", { &temp() }, ());
   |                          ---^^^^^---
   |                          |  |    |
   |                          |  |    temporary value is freed at the end of this statement
   |                          |  creates a temporary value which is freed while still in use
   |                          borrow later used here
   |
   = note: consider using a `let` binding to create a longer lived value

error[E0716]: temporary value dropped while borrowed
  --> $DIR/format-args-temporary-scopes.rs:25:41
   |
LL |     println!("{:?}{:?}", (), if true { &format!("") } else { "" });
   |                                        -^^^^^^^^^^-
   |                                        ||         |
   |                                        ||         temporary value is freed at the end of this statement
   |                                        |creates a temporary value which is freed while still in use
   |                                        borrow later used here
   |
   = note: consider using a `let` binding to create a longer lived value
   = note: this error originates in the macro `format` (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0716`.
